cranelift/codegen/src/isa/x64/lower.isle

```diff
@@ -1521,8 +1521,316 @@
+        (value_regs (add $I64 lo_count hi_count) (imm $I64 0))))
+
+(rule (lower
+       (has_type (ty_32_or_64 ty)
+                 (popcnt src)))
+      (do_popcnt ty src))
+
+(rule (lower
+       (has_type (ty_8_or_16 ty)
+                 (popcnt src)))
+      (do_popcnt $I32 (extend_to_gpr src $I32 (ExtendKind.Zero))))
+
+(rule (lower
+       (has_type $I128
+                 (popcnt src)))
+      (let ((lo_count Gpr (do_popcnt $I64 (value_regs_get_gpr src 0)))
+            (hi_count Gpr (do_popcnt $I64 (value_regs_get_gpr src 1))))
+        (value_regs (add $I64 lo_count hi_count) (imm $I64 0))))
+
+(decl do_popcnt (Type Gpr) Gpr)
+(rule (do_popcnt $I64 src)
+      (let ((shifted1 Gpr (shr $I64 src (Imm8Reg.Imm8 1)))
+            (sevens Gpr (imm $I64 0x7777777777777777))
+            (masked1 Gpr (x64_and $I64 shifted1 sevens))
+            (diff1 Gpr (sub $I64 src masked1))
+            (shifted2 Gpr (shr $I64 masked1 (Imm8Reg.Imm8 1)))
+            (masked2 Gpr (x64_and $I64 shifted2 sevens))
+            (diff2 Gpr (sub $I64 diff1 masked2))
+            (shifted3 Gpr (shr $I64 masked2 (Imm8Reg.Imm8 1)))
+            (masked3 Gpr (x64_and $I64 shifted3 sevens))
+            (diff3 Gpr (sub $I64 diff2 masked3))
+            (sum1 Gpr (add $I64
+                           (shr $I64 diff3 (Imm8Reg.Imm8 4))
+                           diff3))
+            (ofof Gpr (imm $I64 0x0f0f0f0f0f0f0f0f))
+            (masked4 Gpr (x64_and $I64 sum1 ofof))
+            (ones Gpr (imm $I64 0x0101010101010101))
+            (mul Gpr (mul $I64 masked4 ones))
+            (final Gpr (shr $I64 mul (Imm8Reg.Imm8 56))))
+        final))
+(rule (do_popcnt $I32 src)
+      (let ((shifted1 Gpr (shr $I32 src (Imm8Reg.Imm8 1)))
+            (sevens Gpr (imm $I32 0x77777777))
+            (masked1 Gpr (x64_and $I32 shifted1 sevens))
+            (diff1 Gpr (sub $I32 src masked1))
+            (shifted2 Gpr (shr $I32 masked1 (Imm8Reg.Imm8 1)))
+            (masked2 Gpr (x64_and $I32 shifted2 sevens))
+            (diff2 Gpr (sub $I32 diff1 masked2))
+            (shifted3 Gpr (shr $I32 masked2 (Imm8Reg.Imm8 1)))
+            (masked3 Gpr (x64_and $I32 shifted3 sevens))
+            (diff3 Gpr (sub $I32 diff2 masked3))
+            (sum1 Gpr (add $I32
+                           (shr $I32 diff3 (Imm8Reg.Imm8 4))
+                           diff3))
+            (masked4 Gpr (x64_and $I32 sum1 (RegMemImm.Imm 0x0f0f0f0f)))
+            (mul Gpr (mul $I32 masked4 (RegMemImm.Imm 0x01010101)))
+            (final Gpr (shr $I32 mul (Imm8Reg.Imm8 24))))
+        final))
+                       
+            
+(rule 1 (lower (has_type (and
+                          $I8X16
+                          (avx512vl_enabled)
+                          (avx512bitalg_enabled))
+                         (popcnt src)))
+      (vpopcntb src))
+
+
+      
+;; For SIMD 4.4 we use Mula's algorithm (https://arxiv.org/pdf/1611.07612.pdf)
```

Comment:
```suggestion
;; For SSE4.2 we use Mula's algorithm (https://arxiv.org/pdf/1611.07612.pdf)
```
?

Reply:
I had copied this comment from `lower.rs` but it seems that the instructions used only require SSE 4.2 (right?) so I can update accordingly.